### PR TITLE
Eng 6281 - Additional test cases - DML (including UPSERT) with correlated expression subqueries  

### PR DESCRIPTION
**Pull request by alexeevm**

### CODE REVIEW

File: tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java [hunk 1]
```diff
@@ -302,6 +302,59 @@ public void testUpsertWithoutPrimaryKey() throws IOException, ProcCallException
...
+
+            vt = client.callProcedure("@AdHoc", upsert).getResults()[0];
+            validateTableOfLongs(vt, new long[][] {{2}});
+
+            vt = client.callProcedure("@AdHoc", query).getResults()[0];
+            validateTableOfLongs(vt, new long[][] {{1,2,1}, {2,2,2}, {3,3,3}});
+
+            // insert
+            upsert = "UPSERT INTO " + tb + " (ID, WAGE, DEPT) " +
+                    "VALUES((SELECT MAX(ID) + 5 FROM R2), 0, 0);";
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java]: I'm especially interested in what would happen if the subquery produced a non-scalar here (either multiple rows, multiple columns, or both).

Comment (alexeevm) [inline tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java]: It throws the very same "More than one row returned by a scalar/row subquery" exception that is thrown in a similar circumstances with scalar expressions. I will add the negative test case to the suite.

File: tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java [hunk 2]
```diff
@@ -389,12 +442,12 @@ public TestSqlUpsertSuite(String name) {
         assert(success);
         builder.addServerConfig(config);
 
-        //*/ Cluster
-        config = new LocalCluster("sqlupsert-cluster.jar", 2, 3, 1, BackendTarget.NATIVE_EE_JNI);
-        success = config.compile(project);
-        assert(success);
-        builder.addServerConfig(config);
-        //*/
+//        //*/ Cluster
```
Comment (alexeevm) [inline tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java]: Yes, I forgot to uncomment this.

File: tests/frontend/org/voltdb/planner/TestPlansDML.java [hunk 3]
```diff
@@ -282,6 +283,15 @@ public void testDMLwithExpressionSubqueries() {
         dmlSQL = "UPDATE P1 SET C = (SELECT C FROM R2 WHERE A = 0) ;";
         checkDMLPlanNodeAndSubqueryExpression(dmlSQL, null);
 
+        dmlSQL = "UPDATE P1 set C = ? WHERE A = (SELECT MAX(R1.C) FROM R1 WHERE R1.A = P1.A);";
+        checkDMLPlanNodeAndSubqueryExpression(dmlSQL, ExpressionType.COMPARE_EQUAL);
+
+        dmlSQL = "UPDATE P1 set C = (SELECT C FROM R1 WHERE R1.C = P1.C LIMIT 1);";
+        checkDMLPlanNodeAndSubqueryExpression(dmlSQL, null);
+
+        dmlSQL = "UPSERT INTO R6 (A, C) SELECT A, C FROM R2 WHERE NOT EXISTS (SELECT 1 FROM R6 WHERE R6.A = R2.C) ORDER BY 1, 2;";
```
Comment (alexeevm) [inline tests/frontend/org/voltdb/planner/TestPlansDML.java]: There was a blanket restriction on subquery expressions in DML statement regardless of their position in the sql. Once it was removed the subquery expressions in WHERE clause became available.

File: tests/frontend/org/voltdb/regressionsuites/TestSqlDeleteSuite.java [hunk 4]
```diff
@@ -575,6 +575,14 @@ public void testDeleteWithExpresionSubquery()  throws Exception {
             String stmt = "SELECT ID FROM " + table + " ORDER BY ID";
             validateTableOfScalarLongs(client, stmt, new long[] { 4, 5, 6, 7, 8, 9 });
 
+            // delete rows where NUM is 4 and 5
+            vt = client.callProcedure("@AdHoc",
+                    "DELETE FROM " + table + " WHERE NUM IN (SELECT NUM + 2 FROM R1 WHERE R1.ID + 2 = " +
```
Comment (alexeevm) [inline tests/frontend/org/voltdb/regressionsuites/TestSqlDeleteSuite.java]: By the runtime errors you meant the EE errors, right?
Yes, if a scalar expression produces more than a single result we throw a SerializableEEException exception. This is the only exception that can be raised during the scalar expression evaluation. The regular expression (SubqueryExpression::eval) doesn't throw any runtime expressions on its own but obviously propagates any exception raised during the subquery's plan evaluation. The same goes for the scalar expression as well

File: tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java (hunk 1 above)
Comment (alexeevm) [inline tests/frontend/org/voltdb/regressionsuites/TestSqlUpsertSuite.java]: Here is commit with the negative scenarios

https://github.com/VoltDB/voltdb/pull/4396/commits/e5b95fb7a1e7c281e7677ee150fe085970638e20

**Comment by wolffcm** [comment]
Thanks Mike.  I've merged this to the branch in the VoltDB repo so it will run in our automated testing tonight.  I'll let you know how it goes.

**Pull request merged**